### PR TITLE
BUG: Fix off-by-one error in amount of characters in strip

### LINKED ISSUE #26898: BUG: `np.strings.strip` and `np.strings.rstrip` remove unspecified singular remaining characters
### Describe the issue:

`np.strings.strip` and `np.strings.rstrip` are stripping any singular remaining character in addition to the specified char.
For example, `np.strings.rstrip(np.array(['ab', 'ac', 'aab', 'abb']), 'b')` results in `np.array(['', 'ac', 'aa', ''], dtype='<U3')`. I would expect it to result in  `np.array(['a', 'ac', 'aa', 'a'], dtype='<U3')`.

The same behavior was not observed for `lstrip`.

### Reproduce the code example:

```python
import numpy as np
actual = np.strings.rstrip(np.array(['ab', 'ac', 'aab', 'abb']), 'b')
expect = np.array(['a', 'ac', 'aa', 'a'])
np.testing.assert_array_equal(actual, expect)
```

### Error message:

```shell
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[6], line 1
----> 1 np.testing.assert_array_equal(actual, expect)

File ~/mambaforge/envs/mchap_numpy2/lib/python3.11/site-packages/numpy/_utils/__init__.py:85, in _rename_parameter.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
     83             raise TypeError(msg)
     84         kwargs[new_name] = kwargs.pop(old_name)
---> 85 return fun(*args, **kwargs)

    [... skipping hidden 1 frame]

File ~/mambaforge/envs/mchap_numpy2/lib/python3.11/contextlib.py:81, in ContextDecorator.__call__.<locals>.inner(*args, **kwds)
     78 @wraps(func)
     79 def inner(*args, **kwds):
     80     with self._recreate_cm():
---> 81         return func(*args, **kwds)

File ~/mambaforge/envs/mchap_numpy2/lib/python3.11/site-packages/numpy/testing/_private/utils.py:885, in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf, strict, names)
    880         err_msg += '\n' + '\n'.join(remarks)
    881         msg = build_err_msg([ox, oy], err_msg,
    882                             verbose=verbose, header=header,
    883                             names=names,
    884                             precision=precision)
--> 885         raise AssertionError(msg)
    886 except ValueError:
    887     import traceback

AssertionError:
Arrays are not equal

Mismatched elements: 2 / 4 (50%)
 ACTUAL: array(['', 'ac', 'aa', ''], dtype='<U3')
 DESIRED: array(['a', 'ac', 'aa', 'a'], dtype='<U2')
```

### Python and NumPy Versions:

Python: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) [GCC 12.3.0]
NumPy: 2.0.0

### Runtime Environment:

WARNING: `threadpoolctl` not found in system! Install it by `pip install threadpoolctl`. Once installed, try `np.show_runtime` again for more detailed build information
[{'numpy_version': '2.0.0',
  'python': '3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:36:13) '
            '[GCC 12.3.0]',
  'uname': uname_result(system='Linux', node='LIN-L', release='5.10.16.3-microsoft-standard-WSL2', version='#1 SMP Fri Apr 2 22:23:49 UTC 2021', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2',
                                'AVX512F',
                                'AVX512CD',
                                'AVX512_SKX',
                                'AVX512_CLX',
                                'AVX512_CNL',
                                'AVX512_ICL'],
                      'not_found': ['AVX512_KNL', 'AVX512_KNM']}}]

### Context for the issue:

This behavior appears to be incorrect and differs from `np.char.rstrip` in Numpy 1.x.

### PR DESCRIPTION
**Pull request by lysnikolaou**
Closes #26898.

### CODE REVIEW

**Review by seberg — APPROVED** [review]
LGTM, will put it in, thanks.  I found the code slightly hard to follow, I think `i` and `j` could have a clearer name, and suspect making it `len1` rather than `len1 - 1` would be nicer (the way it is `j` can become negative, which works out OK but seemed surprising).

But, I guess it doesn't matter much, so going to put the bug fix in either way.

**Comment by ngoldbaum** [comment]
It looks like this broke something in stringdtype on the asan tests. I'll try to take a look.

**Comment by seberg** [comment]
Thanks, sorry for missing that there was momre than the openblas test failing.

**Comment by ngoldbaum** [comment]
Unfortunately apple clang doesn't trigger it, might need linux or a newer clang release.

**Comment by seberg** [comment]
I think the code is just wrong: It needs a `do {} while ();` rather than that `if` branch it used to have.  That avoids a `--` down too 0 length, I think, which can't work for utf8 I guess.
